### ISSUE TITLE
Cannot run tests in parallel because TypeLocator is using a static cache shared between tests

### ISSUE THREAD

**Issue by idoban** [json-api-dotnet/JsonApiDotNetCore#485]
## Description
Cannot run concurrent tests using TestServer when using Auto Discovery to register types.

Staetup.cs
services.AddJsonApi(facade => facade.AddCurrentAssembly());

System.InvalidOperationException : Collection was modified; enumeration operation may not execute.
Stack Trace:
at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
at JsonApiDotNetCore.Graph.ServiceDiscoveryFacade.AddAssembly(Assembly assembly) in jsonapidotnetcore\src\JsonApiDotNetCore\Graph\ServiceDiscoveryFacade.cs:line 76

The root cause is due to 
private static Dictionary<Assembly, List<ResourceDescriptor>> _identifiableTypeCache = new Dictionary<Assembly, List<ResourceDescriptor>>();

which is being concurrently updated by different tests (each starting up its own TestServer for tests isolation).

see: https://github.com/json-api-dotnet/JsonApiDotNetCore/blob/84f45a93e0a5764a4bde00344699f41325fadc6e/src/JsonApiDotNetCore/Graph/TypeLocator.cs#L15

## Environment

- JsonApiDotNetCore Version: 3.1.0
- Other Relevant Package Versions:

**Comment by maurei** [comment]
I agree it's not optimal to not be able to run the test concurrently, but they still run pretty quick so it's not really bothering me, therefore I would say this is a pretty low priority issue. Thoughts?

**Comment by jaredcnance** [comment]
Can you patch it to use `ConcurrentDictionary` and see if it solves your problem. If it does, then you can open a PR.

**Comment by bart-degreed** [comment]
@idoban Can you verify if #700 works for you?
